### ISSUE TITLE
Wakekolck 0.5.6 not working for some users

### ISSUE THREAD

**Issue by osoftware** [creativecreatorormaybenot/wakelock#153]
Hi!

I'm struggling with finding out why some of the users of my app started reporting that keeping the screen on stopped working after last upgrade. I migrated my legacy native android app which was using FLAG_KEEP_SCREEN_ON to a Flutter based solution using your package (which also just uses FLAG_KEEP_SCREEN_ON, right?). I also bumped compileSdkVersion and targetSdkVersion to 30.

This is an issue that I wasn't able to reproduce on my test device (Redmi 4X, SDK 25), neither on AVDs on various API levels. Examples of devices of users reporting the issue: Redmi 9 (SDK 29), Samsung Galaxy J5 Pro (SDK 27), Samsung Galaxy Note8 (SDK 28), Samsung Galaxy A41 (SDK 30). So it doesn't seem to be related to any particular Android version.

Any idea how to find out what's going on?

**Comment by creativecreatorormaybenot** [comment]
Hi @osoftware - thanks for reporting the issue :)

I am not sure how you can find it out easily.

I have two suggestions:
1. Try to call `Wakelock.enable()` more often; maybe the wakelock is released by some external factor?
2. Check in the code if you can find anything that looks off / could be causing this: https://github.com/creativecreatorormaybenot/wakelock/blob/0a6e8c53125847db31a0494a1662c1bcf719fba8/wakelock/android/src/main/kotlin/creativemaybeno/wakelock/Wakelock.kt#L24

**Comment by osoftware** [comment]
Ok, I've found a reason.

So I was calling `Wakelock.enable()` in `initState` and `Wakelock.disable()` in `dispose` in my page widget. This works fine as long as I navigate between pages with wakelock using `Navigator.push` (as the previous page is not disposed) or leaving them using `Navigator.pop` (as the `dispose` is called as intended). The problem is, when I navigate from a page with wakelock to another page also with wakelock, using `Navigator.pushReplacement`. In that case the previous page is disposed but that happens after the new one is initialized. So the wakelock is disabled by the previus page in the middle of new page lifetime.

To overcome this I used `ReadWriteMutex` from [mutex](https://pub.dev/packages/mutex) package. In `initState` i acquire a read lock (there can be many locks acquired by many pages) then call `Wakelock.enable()`. In `dispose` I release the lock then call `Wakelock.disable()` but only if there is no other lock acquired.

I distilled it into a mixin, maybe someone will find it useful.

```dart

import 'package:flutter/widgets.dart';
import 'package:mutex/mutex.dart';
import 'package:wakelock/wakelock.dart';

final _mutex = ReadWriteMutex();

mixin WakelockMixin<T extends StatefulWidget> on State<T> {
  @override
  void initState() {
    super.initState();

    _mutex.acquireRead();
    Wakelock.enable();
  }

  @override
  void dispose() {
    super.dispose();

    _mutex.release();
    if (_mutex.isLocked) return;
    Wakelock.disable();
  }
}
```
